lib/tasks/stack_check.rake

```diff
@@ -0,0 +1,14 @@
+namespace :stacks do
+  desc 'Check Installed Stacks'
+  task stack_check: :environment do
+    logger = Steno.logger('cc.stack')
+    VCAP::CloudController::Encryptor.db_encryption_key = RakeConfig.config.get(:db_encryption_key)
```

Comment:
**issue/question** Not entirely sure why we do this, but I *believe* `VCAP::CloudController::Encryptor` has two methods of loading an encryption key. There is the historic way which is setting `db_encryption_key` but there is also an array of encryption keys that allow rotation. See [capi-release](https://github.com/cloudfoundry/capi-release/blob/develop/jobs/cloud_controller_ng/spec#L766-L781). You can see this logic here https://github.com/cloudfoundry/cloud_controller_ng/blob/3d09a707b253acbdf54dff659a329f4267c840f3/lib/cloud_controller/encryptor.rb#L49-L51. Here we seem to only set the old style encryption key, so this does not factor in the encryption array/ledger.

So I'm not sure what the consequence is of loading this Encrypter this way. Do we need to create it with more configuration, somewhat like https://github.com/cloudfoundry/cloud_controller_ng/blob/main/lib/cloud_controller/config.rb#L110-L116 ?

Reply:
Not entirely sure how we use `config.rb`, we load the encryption key the same way as db migrations,
https://github.com/cloudfoundry/cloud_controller_ng/blob/3d09a707b253acbdf54dff659a329f4267c840f3/lib/tasks/db.rake#L127

@Gerg @tcdowney any input here? is there a correct or preferred way to load the db encryption key?

Reply:
I think @Samze is correct and I think maybe we've found a bug in `db.rake` -- unless something somewhere is calling `@config.configure_components` and it gets set up properly as a side effect.

I think both here and `db.rake` need to do the full thing and load the encryption key all the ways in these rake tasks (maybe this could be put into a function on the config object).

```
Encryptor.db_encryption_key = get(:db_encryption_key)

      if get(:database_encryption)
        Encryptor.database_encryption_keys = get(:database_encryption)[:keys]
        Encryptor.current_encryption_key_label = get(:database_encryption)[:current_key_label]
        Encryptor.pbkdf2_hmac_iterations = get(:database_encryption)[:pbkdf2_hmac_iterations]
      end
```

The legacy way of using `db_encryption_key` works when that's set, but it's a deprecated field and the array of properties is the preferred way of doing this.

https://github.com/cloudfoundry/cf-deployment/blob/2893dec3cb0afaab310f8f23b235969f21c9cd28/cf-deployment.yml#L946-L950

I think this might be able to cause some subtle bugs with the way it currently is... 🤔

Reply:
It does the same thing again here:
https://github.com/cloudfoundry/cloud_controller_ng/blob/3d09a707b253acbdf54dff659a329f4267c840f3/lib/cloud_controller/db_migrator.rb#L8

🤔

Reply:
Yeah I put a `binding.pry` in the DBMigrator and confirmed the proper one is empty:
```
[4] pry(#<DBMigrator>)> VCAP::CloudController::Encryptor.send(:database_encryption_keys)
=> {}
```

I think we might want to make a separate bug and fix that @Samze @xandroc ^ Good catch

Reply:
I'll write a story for this, thanks for verifying @tcdowney

Reply:
@Samze updated

Reply:
@xandroc I don't know much about whether just calling `@config.configure_components` here has any additional consequences or if it is safe, perhaps @tcdowney knows? Just want to make sure we're not introducing additional problems here.

Reply:
I don't know for sure. The `ProcessObserver` setup and `run_initializers` might not be something we want to do, but I'm not positive.

https://github.com/cloudfoundry/cloud_controller_ng/blob/3d09a707b253acbdf54dff659a329f4267c840f3/lib/cloud_controller/config.rb#L121-L123

To be safe you might want to extract the db encryptor steps out into its own method and call that in configure_components and these rake tasks.

Reply:
Broke db setup out into to its own method and calling it separately